Mathlib/RingTheory/LocalRing/Module.lean

```diff
@@ -0,0 +1,264 @@
+  · rintro rfl; simp
+
+theorem LocalRing.map_mkQ_eq {N₁ N₂ : Submodule R M} (h : N₁ ≤ N₂) (h' : N₂.FG) :
+    N₁.map (Submodule.mkQ (𝔪 • N₂)) = N₂.map (Submodule.mkQ (𝔪 • N₂)) ↔ N₁ = N₂ := by
+  constructor
+  · intro hN
+    have : N₂ ≤ 𝔪 • N₂ ⊔ N₁ := by
+      simpa using Submodule.comap_mono (f := Submodule.mkQ (𝔪 • N₂)) hN.ge
+    rw [sup_comm] at this
+    exact h.antisymm (Submodule.le_of_le_smul_of_le_jacobson_bot h'
+      (by rw [jacobson_eq_maximalIdeal]; exact bot_ne_top) this)
+  · rintro rfl; simp
+
+theorem LocalRing.map_mk_eq_top {N : Submodule R M} [Module.Finite R M] :
+    N.map (TensorProduct.mk R k M 1) = ⊤ ↔ N = ⊤ := by
+  constructor
+  · intro hN
+    letI : Module k (M ⧸ (𝔪 • ⊤ : Submodule R M)) :=
+      inferInstanceAs (Module (R ⧸ 𝔪) (M ⧸ 𝔪 • (⊤ : Submodule R M)))
+    letI : IsScalarTower R k (M ⧸ (𝔪 • ⊤ : Submodule R M)) :=
+      inferInstanceAs (IsScalarTower R (R ⧸ 𝔪) (M ⧸ 𝔪 • (⊤ : Submodule R M)))
+    let f := AlgebraTensorModule.lift (((LinearMap.ringLmapEquivSelf k k _).symm
+      (Submodule.mkQ (𝔪 • ⊤ : Submodule R M))).restrictScalars R)
+    have : f.comp (TensorProduct.mk R k M 1) = Submodule.mkQ (𝔪 • ⊤) := by ext; simp [f]
+    have hf : Function.Surjective f := by
+      intro x; obtain ⟨x, rfl⟩ := Submodule.mkQ_surjective _ x;
+      rw [← this, LinearMap.comp_apply]; exact ⟨_, rfl⟩
+    apply_fun Submodule.map f at hN
+    rwa [← Submodule.map_comp, this, Submodule.map_top, LinearMap.range_eq_top.mpr hf,
+      LocalRing.map_mkQ_eq_top] at hN
+  · rintro rfl; rw [Submodule.map_top, LinearMap.range_eq_top]
+    exact TensorProduct.mk_surjective R M k Ideal.Quotient.mk_surjective
+
+theorem LocalRing.subsingleton_tensorProduct [Module.Finite R M] :
+    Subsingleton (k ⊗[R] M) ↔ Subsingleton M := by
+  rw [← Submodule.subsingleton_iff R, ← subsingleton_iff_bot_eq_top,
+    ← Submodule.subsingleton_iff R, ← subsingleton_iff_bot_eq_top,
+    ← LocalRing.map_mk_eq_top (M := M), Submodule.map_bot]
+
+theorem LocalRing.span_eq_top_of_tmul_eq_basis [Module.Finite R M] {ι}
+    (f : ι → M) (b : Basis ι k (k ⊗[R] M))
+    (hb : ∀ i, 1 ⊗ₜ f i = b i) : Submodule.span R (Set.range f) = ⊤ := by
+  rw [← LocalRing.map_mk_eq_top, Submodule.map_span, ← Submodule.restrictScalars_span R k
+    Ideal.Quotient.mk_surjective, Submodule.restrictScalars_eq_top_iff,
+    ← b.span_eq, ← Set.range_comp]
+  simp only [Function.comp, mk_apply, hb, Basis.span_eq]
+
+open Function in
+/--
+Given `M₁ → M₂ → M₃ → 0` and `N₁ → N₂ → N₃ → 0`.
+If `M₁ ⊗ N₃ → M₂ ⊗ N₃` and `M₂ ⊗ N₁ → M₂ ⊗ N₂` are both injective,
+then `M₃ ⊗ N₁ → M₃ ⊗ N₂` is also injective.
+-/
+theorem lTensor_injective_of_exact_of_exact_of_rTensor_injective
+    {M₁ M₂ M₃ N₁ N₂ N₃}
+    [AddCommGroup M₁] [Module R M₁] [AddCommGroup M₂] [Module R M₂] [AddCommGroup M₃] [Module R M₃]
+    [AddCommGroup N₁] [Module R N₁] [AddCommGroup N₂] [Module R N₂] [AddCommGroup N₃] [Module R N₃]
+    (f₁ : M₁ →ₗ[R] M₂) (f₂ : M₂ →ₗ[R] M₃) (g₁ : N₁ →ₗ[R] N₂) (g₂ : N₂ →ₗ[R] N₃)
```

Comment:
Why not leaving `f₁, f₂, g₁, g₂` implicit? I tried in the application below and it works.